File: Sources/ExcalidrawMath/Polygon.swift
import Foundation

/// A closed polygon (`packages/math/src/polygon.ts`). The stored `points` are
/// closed: the first point is repeated as the last when needed.
public struct Polygon: Equatable, Sendable {
    public let points: [Point]

    public init(_ points: [Point]) {
        self.points = Polygon.closed(points)
    }

    private static func closed(_ points: [Point]) -> [Point] {
        guard let first = points.first, let last = points.last else { return points }
        return first.isApproximatelyEqual(to: last) ? points : points + [first]
    }

    /// Even-odd ray-casting containment test (`polygonIncludesPoint`).
    public func includes(_ point: Point) -> Bool {
        let x = point.x
        let y = point.y
        var inside = false
        var j = points.count - 1
        for i in 0..<points.count {
            let xi = points[i].x, yi = points[i].y
            let xj = points[j].x, yj = points[j].y
            if ((yi > y && yj <= y) || (yi <= y && yj > y)),
               x < (xj - xi) * (y - yi) / (yj - yi) + xi {
                inside.toggle()
            }
            j = i
        }
        return inside
    }

    /// Non-zero winding-number containment test (`polygonIncludesPointNonZero`).
    public func includesNonZero(_ point: Point) -> Bool {
        let x = point.x
        let y = point.y
        var winding = 0
        for i in 0..<points.count {
            let j = (i + 1) % points.count
            let xi = points[i].x, yi = points[i].y
            let xj = points[j].x, yj = points[j].y
            if yi <= y {
                if yj > y, (xj - xi) * (y - yi) - (x - xi) * (yj - yi) > 0 {
                    winding += 1
                }
            } else if yj <= y, (xj - xi) * (y - yi) - (x - xi) * (yj - yi) < 0 {
                winding -= 1
            }
        }
        return winding != 0
    }

    /// Whether `point` lies on any edge within `threshold` (`pointOnPolygon`).
    public func contains(_ point: Point, threshold: Double = ExcalidrawMath.precision) -> Bool {
        guard points.count >= 2 else { return false }
        for i in 0..<(points.count - 1) where
            LineSegment(points[i], points[i + 1]).contains(point, threshold: threshold) {
            return true
        }
        return false
    }
}

File: web/packages/math/src/polygon.ts
import { PRECISION, type Point } from "./point.js";
import { LineSegment } from "./segment.js";

/**
 * A closed polygon. The stored `points` are closed: the first point is repeated
 * as the last when needed. (parity: Polygon.swift)
 */
export class Polygon {
  readonly points: Point[];

  constructor(points: Point[]) {
    this.points = Polygon.closed(points);
  }

  private static closed(points: Point[]): Point[] {
    const first = points[0];
    const last = points[points.length - 1];
    if (first === undefined || last === undefined) return points;
    return first.isApproximatelyEqual(last) ? points : [...points, first];
  }

  /** Even-odd ray-casting containment test (`polygonIncludesPoint`). */
  includes(point: Point): boolean {
    const { x, y } = point;
    let inside = false;
    let j = this.points.length - 1;
    for (let i = 0; i < this.points.length; i++) {
      const pi = this.points[i]!;
      const pj = this.points[j]!;
      const xi = pi.x;
      const yi = pi.y;
      const xj = pj.x;
      const yj = pj.y;
      if (
        ((yi > y && yj <= y) || (yi <= y && yj > y)) &&
        x < ((xj - xi) * (y - yi)) / (yj - yi) + xi
      ) {
        inside = !inside;
      }
      j = i;
    }
    return inside;
  }

  /** Non-zero winding-number containment test (`polygonIncludesPointNonZero`). */
  includesNonZero(point: Point): boolean {
    const { x, y } = point;
    let winding = 0;
    for (let i = 0; i < this.points.length; i++) {
      const jdx = (i + 1) % this.points.length;
      const pi = this.points[i]!;
      const pj = this.points[jdx]!;
      const xi = pi.x;
      const yi = pi.y;
      const xj = pj.x;
      const yj = pj.y;
      if (yi <= y) {
        if (yj > y && (xj - xi) * (y - yi) - (x - xi) * (yj - yi) > 0) {
          winding += 1;
        }
      } else if (yj <= y && (xj - xi) * (y - yi) - (x - xi) * (yj - yi) < 0) {
        winding -= 1;
      }
    }
    return winding !== 0;
  }

  /** Whether `point` lies on any edge within `threshold` (`pointOnPolygon`). */
  contains(point: Point, threshold = PRECISION): boolean {
    if (this.points.length < 2) return false;
    for (let i = 0; i < this.points.length - 1; i++) {
      if (new LineSegment(this.points[i]!, this.points[i + 1]!).contains(point, threshold)) {
        return true;
      }
    }
    return false;
  }
}
